### DISCUSSION TITLE
How to hack around iframe components?

### DISCUSSION THREAD

**Discussion by renhiyama** [servo/servo#28914] [General]
I was thinking of making a headless app with servo with the motive of making a os mockup (or possibly a internet os, as described in [Wikipedia](https://en.m.wikipedia.org/wiki/Internet_OS#:~:text=An%20Internet%20operating%20system%2C%20or,client%2C%20generally%20a%20web%20browser.)

And to achieve low resource usage, i would like to use iframes for managing "apps/websites" instead of multiple instances. And iframes does sound better if I'm making a mockup of a app's window inside the web app itself. But the problem is, most of the websites would not allow loading in iframes (due to cors and stuff) so i would like to make a new iframe like component for my app itself, which doesn't follow cors, and not allow the same custom iframe component to be ran inside itself, to avoid other sites misusing this feature.

Therefore, i want to know where is the code for the iframe element in this repo (sorry but i couldn't find it) and wanted to know where is it 😅.

**Comment by yvt** [comment]
Hi @renhiyama,

Sorry for only being able to give an abstract answer, but being an instrument to connect between potentially-cross-process browsing contexts, the implementation of iframes is extremely convoluted and spread across multiple components (especially `script` and `constellation`). The module [`script::dom::htmliframeelement`](https://github.com/servo/servo/blob/c23adde5a33348aa38f7c2ad22bf96d91dcc7b33/components/script/dom/htmliframeelement.rs#L233) contains the API implementation as well as some of the entry logic relevant to the creation and destruction of iframes, so you might want to start looking from there and follow function calls and IPC messages.

Another good idea is to look for where exactly the restriction (`X-Frame-Options` in this case) is imposed, and then find out how to connect it to an API that a web content can interact with.
